continuous_integration/scripts/test_report.py

```diff
@@ -176,58 +233,89 @@ def dedup(group):
         for w in workflows
         if (
             pandas.to_datetime(w["created_at"])
-            > pandas.Timestamp.now(tz="UTC") - pandas.Timedelta(days=90)
+            > pandas.Timestamp.now(tz="UTC") - pandas.Timedelta(days=days)
             and w["conclusion"] != "cancelled"
             and w["name"].lower() == "tests"
         )
     ]
+    print(f"Found {len(workflows)} workflows")
     # Each workflow processed takes ~10-15 API requests. To avoid being
     # rate limited by GitHub (1000 requests per hour) we choose just the
     # most recent N runs. This also keeps the viz size from blowing up.
-    workflows = sorted(workflows, key=lambda w: w["created_at"])[-50:]
+    workflows = sorted(workflows, key=lambda w: w["created_at"])[-max_workflows:]
+    print(f"Fetching artifact listing for the {len(workflows)} most recent workflows")
 
-    print("Getting the artifact listing for each workflow...")
     for w in workflows:
-        artifacts = get_artifacts_for_workflow(w["id"])
+        artifacts = get_artifacts_for_workflow(w["id"], repo=repo)
         # We also upload timeout reports as artifacts, but we don't want them here.
         w["artifacts"] = [
             a
             for a in artifacts
             if "timeouts" not in a["name"] and "cluster_dumps" not in a["name"]
         ]
 
-    print("Downloading and parsing artifacts...")
-    for w in workflows:
-        w["dfs"] = []
-        for a in w["artifacts"]:
-            xml = download_and_parse_artifact(a["archive_download_url"])
-            df = dataframe_from_jxml(xml) if xml else None
-            # Note: we assign a column with the workflow timestamp rather than the
-            # artifact timestamp so that artifacts triggered under the same workflow
-            # can be aligned according to the same trigger time.
-            if df is not None:
-                df = df.assign(
-                    name=a["name"],
-                    suite=suite_from_name(a["name"]),
-                    date=w["created_at"],
-                    url=w["html_url"],
-                )
-                w["dfs"].append(df)
-
-    # Make a top-level dict of dataframes, mapping test name to a dataframe
-    # of all check suites that ran that test.
-    # Note: we drop **all** tests which did not have at least one failure.
+    nartifacts = sum(len(w["artifacts"]) for w in workflows)
+    ndownloaded = 0
+    print(f"Downloading and parsing {nartifacts} artifacts...")
+
+    with shelve.open("test_report") as cache:
```

Comment:
> I guess this should do a good job of ameliorating rate limiting as well.

The shelf is on a local file which is lost every time you restart the test, so no.
Storing the cache somewhere more permanent would let us show the full 90 days instead of just ~20, but
- it would also require some sort of expiration logic
- regenerating the cache from scratch (for whatever reason) on the full 120 days would be challenging, as it would hit the rate limiter.

>  We may still start running into rate limiting issues since this doesn't cache everything

Yes, we're downloading the list of workflows and artifacts twice now. I hope that's not the tipping point...

Reply:
> Yes, we're downloading the list of workflows and artifacts twice now. I hope that's not the tipping point...

Ooh, this might be an issue then, we were fairly close to the rate limit as is.

Reply:
We could reduce the number of days from 90 to 21? Currently we're downloading 17 days due to the limit of 50 workflows.